File: CyScat/Scattering_Code/farefieldall.jl
"""
Farfield Calculator
Coder : Curtis Jin
Date  : 2010/OCT/21st Thursday
Contact : jsirius@umich.edu
Description : Professor Michelson's version
            : Far field calculator

Julia translation
"""

"""
    farefieldall(clocs, cmmaxs, lambda_val, coefficients, nphi)

Calculate the far field for all cylinders.

# Arguments
- `clocs`: Cylinder locations (ncyl x 2)
- `cmmaxs`: Maximum mode numbers for each cylinder
- `lambda_val`: Wavelength
- `coefficients`: Scattering coefficients
- `nphi`: Number of angular points

# Returns
- `result`: Far field values with angles (nphi x 2), first column is angles, second is field
"""
function farefieldall(clocs, cmmaxs, lambda_val, coefficients, nphi)
    efieldlist = zeros(ComplexF64, nphi)
    nstart = 1  # Julia 1-indexed

    for icyl in 1:length(cmmaxs)
        cmmax = Int(cmmaxs[icyl])
        num_coeffs = 2 * cmmax + 1
        coeff_slice = coefficients[nstart:nstart + num_coeffs - 1]

        efieldlist .+= farefieldone(
            clocs[icyl, :], cmmax, lambda_val, coeff_slice, nphi
        )
        nstart = nstart + num_coeffs
    end

    # Create angle list and combine with field values
    philist = collect(0:nphi-1) .* 2π / nphi

    # Return as (nphi, 2) array with angles and field values
    result = hcat(philist, efieldlist)

    return result
end


"""
    farefieldone(cloc, cmmax, lambda_val, coefficient, nphi)

Calculate the far field for a single cylinder.
"""
function farefieldone(cloc, cmmax, lambda_val, coefficient, nphi)
    efieldlist = zeros(ComplexF64, nphi)

    for cm in -cmmax:cmmax
        idx = cm + cmmax + 1  # Julia 1-indexed
        efieldlist .+= farefieldonem(
            cloc, cm, lambda_val, coefficient[idx], nphi
        )
    end

    return efieldlist
end


"""
    farefieldonem(cloc, cm, lambda_val, coefficientm, nphi)

Calculate the far field for a single mode of a single cylinder.
"""
function farefieldonem(cloc, cm, lambda_val, coefficientm, nphi)
    dphi = 2π / nphi
    k = 2π / lambda_val

    efieldlist = zeros(ComplexF64, nphi)
    for iphi in 1:nphi
        phi = (iphi - 1) * dphi
        phase1 = exp(1im * k * (cloc[1] * cos(phi) + cloc[2] * sin(phi)))
        phase2 = exp(1im * cm * phi)
        amplitude = sqrt(2 / (π * k))
        phase3 = exp(1im * (cm * π / 2 + π / 4))
        efieldlist[iphi] = phase1 * phase2 * amplitude * phase3 * coefficientm
    end

    return efieldlist
end

File: python_jax/CyScat/Scattering_Code/farefieldall.py
"""
Farfield Calculator
Coder : Curtis Jin
Date  : 2010/OCT/21st Thursday
Contact : jsirius@umich.edu
Description : Professor Michelson's version
            : Far field calculator

Python translation
"""

import numpy as np


def farefieldall(clocs, cmmaxs, lambda_val, coefficients, nphi):
    """
    Calculate the far field for all cylinders.

    Parameters:
    -----------
    clocs : ndarray
        Cylinder locations (ncyl x 2)
    cmmaxs : ndarray
        Maximum mode numbers for each cylinder
    lambda_val : float
        Wavelength
    coefficients : ndarray
        Scattering coefficients
    nphi : int
        Number of angular points

    Returns:
    --------
    efieldlist : ndarray
        Far field values with angles (nphi x 2), first column is angles, second is field
    """
    efieldlist = np.zeros(nphi, dtype=complex)
    nstart = 0  # 0-indexed in Python

    for icyl in range(len(cmmaxs)):
        cmmax = cmmaxs[icyl]
        num_coeffs = 2 * cmmax + 1
        coeff_slice = coefficients[nstart:nstart + num_coeffs]

        efieldlist = efieldlist + farefieldone(
            clocs[icyl, :], cmmax, lambda_val, coeff_slice, nphi
        )
        nstart = nstart + num_coeffs

    # Create angle list and combine with field values
    philist = np.arange(0, nphi) * 2 * np.pi / nphi

    # Return as (nphi, 2) array with angles and field values
    result = np.column_stack([philist, efieldlist])

    return result


def farefieldone(cloc, cmmax, lambda_val, coefficient, nphi):
    """
    Calculate the far field for a single cylinder.
    """
    efieldlist = np.zeros(nphi, dtype=complex)

    for cm in range(-cmmax, cmmax + 1):
        idx = cm + cmmax  # coefficient index
        efieldlist = efieldlist + farefieldonem(
            cloc, cm, lambda_val, coefficient[idx], nphi
        )

    return efieldlist


def farefieldonem(cloc, cm, lambda_val, coefficientm, nphi):
    """
    Calculate the far field for a single mode of a single cylinder.
    """
    dphi = 2 * np.pi / nphi
    k = 2 * np.pi / lambda_val

    efieldlist = np.zeros(nphi, dtype=complex)
    for iphi in range(nphi):
        phi = iphi * dphi
        # exp(1i*k*cloc*[cos(phi);sin(phi)]) * exp(1i*cm*phi) * sqrt(2/(pi*k)) * exp(1i*(cm*pi/2+pi/4)) * coefficientm
        phase1 = np.exp(1j * k * (cloc[0] * np.cos(phi) + cloc[1] * np.sin(phi)))
        phase2 = np.exp(1j * cm * phi)
        amplitude = np.sqrt(2 / (np.pi * k))
        phase3 = np.exp(1j * (cm * np.pi / 2 + np.pi / 4))
        efieldlist[iphi] = phase1 * phase2 * amplitude * phase3 * coefficientm

    return efieldlist
